src/notebookConcatDocument.ts

```diff
@@ -261,95 +343,179 @@ export class NotebookConcatDocument implements vscode.TextDocument, vscode.Dispo
-        if (range) {
-            const startPosition = this.positionAt(range.startOffset);
-            const endPosition = this.positionAt(range.endOffset);
-            return new vscode.Range(startPosition, endPosition);
+        return new vscode.Position(0, 0);
+    }
+
+    public concatOffsetAt(location: vscode.Location): number {
+        // Location is inside of a cell
+        const firstRealLine = this._realLines.find((r) => r.cellUri.toString() === location.uri.toString());
+        if (firstRealLine) {
+            // Line number is inside a real line
+            const realLine = this._realLines[location.range.start.line + firstRealLine.lineNumber];
+
+            // Use its offset (real offset) to find our outgoing offset
+            return this.mapRealToConcatOffset(realLine.offset + location.range.start.character);
         }
+        return 0;
     }
-    public getText(range?: vscode.Range | undefined): string {
-        if (!range) {
-            return this._contents;
-        } else {
-            const startOffset = this.convertToOffset(range.start);
-            const endOffset = this.convertToOffset(range.end);
-            return this._contents.substring(startOffset, endOffset - startOffset);
+
+    public concatRangeOf(cellUri: vscode.Uri) {
+        const cellLines = this._lines.filter((l) => l.cellUri.toString() === cellUri.toString());
+        const firstLine = cellLines[0];
+        const lastLine = cellLines[cellLines.length - 1];
+        if (firstLine && lastLine) {
+            return new vscode.Range(firstLine.range.start, lastLine.rangeIncludingLineBreak.end);
         }
+        return new vscode.Range(new vscode.Position(0, 0), new vscode.Position(0, 0));
+    }
+    public realRangeOf(cellUri: vscode.Uri) {
+        // Get all the real spans
+        const realSpans = this._spans.filter((s) => s.uri.toString() == cellUri.toString() && s.inRealCell);
+        const startOffset = realSpans[0].startOffset | 0;
+        const endOffset = realSpans.length > 0 ? realSpans[realSpans.length - 1].endOffset : startOffset;
+
+        // Find the matching concat lines
+        const firstLine = this._lines.find((l) => startOffset >= l.offset && startOffset < l.endOffset);
+        const lastLine = this._lines.find((l) => endOffset >= l.offset && endOffset < l.endOffset);
+        if (firstLine && lastLine) {
+            return new vscode.Range(firstLine.range.start, lastLine.rangeIncludingLineBreak.end);
+        }
+        return new vscode.Range(new vscode.Position(0, 0), new vscode.Position(0, 0));
     }
     public getCells(): vscode.Uri[] {
-        return this._cellRanges.map((c) => c.uri);
+        return [...new Set(this._spans.map((c) => c.uri))];
     }
-    public locationAt(positionOrRange: vscode.Range | vscode.Position): vscode.Location {
+
+    public notebookLocationAt(positionOrRange: vscode.Range | vscode.Position): vscode.Location {
+        // positionOrRange should be in concat ranges
```

Comment:
Minor suggestion: 
const range = (positionOrRange instanceof vscode.Position) ? new vscode.Range(positionOrRange, positionOrRange) : positionOrRange;

Maybe pedantic, but then you could use range for the rest of the function instead of positionOrRange when you are assuring that you have a Range here.

Reply:
Yes that's clearer, will fix